File: 3. Behavioral Design Patterns/Mediator/Typescript/MediatorTemplate.ts
// Mediator Pattern – TypeScript Template

interface Mediator { notify(sender: any, eventType: string): void }

abstract class Colleague {
  constructor(protected mediator: Mediator) {}
  abstract doSomething(): void
}

class ConcreteColleagueA extends Colleague {
  doSomething(): void { this.mediator.notify(this, 'EventA') }
}

class ConcreteColleagueB extends Colleague {
  doSomething(): void { this.mediator.notify(this, 'EventB') }
}

class ConcreteMediator implements Mediator {
  private colleagues: Colleague[] = []
  addColleague(colleague: Colleague): void { this.colleagues.push(colleague) }
  notify(sender: any, eventType: string): void { /* handle event */ }
}

;(function main() {
  const mediator = new ConcreteMediator()
  const colleagueA = new ConcreteColleagueA(mediator)
  const colleagueB = new ConcreteColleagueB(mediator)
  mediator.addColleague(colleagueA)
  mediator.addColleague(colleagueB)
  colleagueA.doSomething()
})()

export {}

File: 3. Behavioral Design Patterns/Mediator/Kotlin/MediatorTemplate.kt
// Mediator Pattern – Kotlin Template

interface Mediator { fun notify(sender: Any, eventType: String) }

abstract class Colleague(protected val mediator: Mediator) {
    abstract fun doSomething()
}

class ConcreteColleagueA(mediator: Mediator) : Colleague(mediator) {
    override fun doSomething() = mediator.notify(this, "EventA")
}

class ConcreteColleagueB(mediator: Mediator) : Colleague(mediator) {
    override fun doSomething() = mediator.notify(this, "EventB")
}

class ConcreteMediator : Mediator {
    private val colleagues = mutableListOf<Colleague>()
    fun addColleague(colleague: Colleague) = colleagues.add(colleague).let {}
    override fun notify(sender: Any, eventType: String) { /* handle event */ }
}

fun main() {
    val mediator = ConcreteMediator()
    val colleagueA = ConcreteColleagueA(mediator)
    val colleagueB = ConcreteColleagueB(mediator)
    mediator.addColleague(colleagueA)
    mediator.addColleague(colleagueB)
    colleagueA.doSomething()
}
